### DISCUSSION TITLE
Editor performance issues in large codebases

### DISCUSSION THREAD

**Discussion by tonyxiao** [ecyrbe/zodios#417] [Q&A]
Using https://github.com/astahmer/openapi-zod-client and example schema https://api.merge.dev/api/accounting/v1/schema, we generated a zodios file that is about 5k lines long. At this length it seems that the autocomplete works only after a very significant lag, making it really difficult to use zodios. Does this happen to anyone else? Are there ways we can make the performance better?

**Comment by ecyrbe** [comment]
Hello,
I'll migrate this into Q&A because we are wrapping the limits of typescript there (not an issue i can address)
So here are some thoughs:
- if you can, remove all the endpoints and schema you don't need on your code
- Split it into different files and different client zodios instance / like put a client for each resources (if you have isolated modules ts perf will go up)

**Reply by tonyxiao** [reply]
The thing is I'm currently using https://github.com/drwpow/openapi-typescript at the moment together with https://github.com/ajaishankar/openapi-typescript-fetch and it performs just fine (even though 15k lines of types were generated).

**Reply by ecyrbe** [reply]
When i say it's typescript limits, it's typescript limits with generics. 
Generics are slowers than raw types and generated code endpoints.
But maybe there are some tricks zodios could do to improve this.
I have some ideas. I'll try them on your example. Maybe i can improve things, but no guaranty for now.

**Reply by ecyrbe** [reply]
@tonyxiao i just tested on zodios v11 (where i got rid of generics narrowing and use ts 5.0 const generics instead) and i actually have good performances.

So maybe switch to zodios beta v11 for now ?

**Reply by ecyrbe** [reply]
I just published the TS 5.0 version so you can try it :
🦋  info Publishing "@zodios/axios" at "11.0.0-beta.18"
🦋  info Publishing "@zodios/core" at "11.0.0-beta.18"
🦋  info Publishing "@zodios/express" at "11.0.0-beta.18"
🦋  info Publishing "@zodios/fetch" at "11.0.0-beta.18"
🦋  info Publishing "@zodios/react" at "11.0.0-beta.18"
🦋  info Publishing "@zodios/testing" at "11.0.0-beta.18"

you'll need to install `@zodios/core` with a client `@zodios/axios` or `@zodios/fetch`. if you use `@zodios/react` client can use any of the fetchers (axios or fetch)

**Reply by tonyxiao** [reply]
Exciting! Will need to update our project to ts5.0 (about time) and circle back to test this out!

**Reply by tonyxiao** [reply]
Also really appreciate that in v11 axios becomes a plugin. There are plenty of situations where we need to handle custom fetch logic, etc. just a short list of platforms I interact with

- Coda
- Pipedream
- Browser
- Server
- Deno / Supabase functions
- Vercel edge function
- Drafts / JavaScriptCore
- React native

By having no hard dependencies we enable zodios to work across all platforms with a single api & common validation 🤯

**Reply by tonyxiao** [reply]
Seems that `new Zodios` doesn't work anymore? `Zodios` is no longer exported from core.

**Reply by tonyxiao** [reply]
Ran into some issues while testing
part 1: https://www.loom.com/share/9f287982d32b4bd5a355c3eeaa284d68
Part 2: https://www.loom.com/share/2a13c68fbee14952964f569139ddd4c8

**Reply by ecyrbe** [reply]
Zodios is now only exported from @zodios/fetch or @zodios/axios

**Reply by ecyrbe** [reply]
The first issue is due to the fact that now the body has to be filled in the body property. There is no more a distinct function parameter for body.

**Reply by ecyrbe** [reply]
The second issue, i can't reproduce. i imported airbyte api and have the return typed correctly.

**Reply by ecyrbe** [reply]
But i have big performances issues with full airbyte api. it's unsuable without splitting defintions. My guess is that the filtering takes too much time. but i'll investigate why typescript is so slow at filtering.

**Reply by tonyxiao** [reply]
> The first issue is due to the fact that now the body has to be filled in the body property. There is no more a distinct function parameter for body.
> 
> 

I tried it but the body property is coming out to be unknown

**Reply by AntoineThibi** [reply]
I had the same issue of editor performance and I tried the version 11.0.0-beta.19. It seems to reduce by about 50% the typescript time. 

However, I had an issue getting to the instance of axios in order to had interceptor.

```
import { Zodios } from "@zodios/axios";
import { makeApi } from "@zodios/core";

const Endpoints = makeApi(myEndpoints);

export const myApi = new Zodios(Endpoints, {
  axiosConfig: {
    baseUrl: 'myurl'
  }
});

// Previously I did : 
myApi.axios.interceptors.request.use(....)

// Now after looking at the typing : 
myApi._fetcherProvider.options.axiosInstance.interceptors.request.use(.....)
// However while running it I find that _fetcherProvider is undefined
```

Do you have any idea were it can come from ?

**Reply by ecyrbe** [reply]
Axios is no more exposed as zodios v11 is now fetcher agnostic. 
Either pass your own instance of axios and plug your interceptors in your custom axios instance or better use zodios plugins they are the same: 
https://www.zodios.org/docs/client/plugins#write-your-own-plugin

**Reply by jordanmnunez** [reply]
@ecyrbe We are migrating to v11-beta to fix these same editor issues and are are seeing solid improvements as well (50% reduction in type compiles). However, we were also borrowing some typing from v10 that are hard to track down now.
For example:  `import type { ZodiosAliases } from "@zodios/core/lib/zodios.types";` and `import { ZodiosError, zodValidationPlugin } from "@zodios/core";`

Can you please share any updated docs or links to source code for v11-beta. I have been unable to track them down.

**Comment by Dimava** [comment]
I strongly think this is caused by Zodios using type-level `array.find({ path })` rather then `record[path]`, leading to `O(m*n)` rather then `O(M + n)`

**Reply by Dimava** [reply]
TS won't recompile record once it's compiled, (so big M), but will iterate array because it has to (so m*n)

**Reply by ecyrbe** [reply]
No, no,
We don't have double loops in find, so complexity is also additive with arrays
The diff with array vs record is O(1) vs O(N)
So we have O(M+N) vs O(N)

**Reply by Dimava** [reply]
I mean, if there are N routes
 - the instantiation of `const Record<route, Route>` will take `O(N)`
   - the search in `Record<route, Route>` will take `O(1)`
     - the search for every route will take `O(N)`
 - the instantiation of `const Route[]` will take `O(N)`
   - the search in `Route[]` will take `O(N)`
     - the search for every route will take `O(N*N)`

**Reply by Dimava** [reply]
ArkType has type instantiation benchmark, which could be used in the future

ArkType's type performance is an order of magnitude better then Zod's exactly because it's made as efficiently as possible.
I think it would make sence to compare type instantiation performance of TRPC/ts-rest/Zodios as they are generally equivalent

**Comment by antonio-ivanovski** [comment]
We were facing severe performance issues for a fairly small application with Zodios v10, so I tried to migrate to v11. Here are some findings:

1. 🟢 Resolutions and autocomplete are almost instantaneous. From the previous ~10 seconds to even timing out and failing to resolve the type, currently, it is <200ms.
2. 🟢 The `isErrorFromPath` and `isErrorFromAlias` has been removed from utility kind to `ZodiosInstance` method
3. 🟢 Making requests is cleaner by having both `body` and `params` properties in the method.
4. 🟢 Server side `ZodiosRequestHandler` is 1 to 1 mapped to `ZodiosPathMiddleware`.
5. 🟡 Missing the `paramtersBuilder` utility in the `beta.19`, resorting back to `makeParameters`.
6. 🟡 Dynamic setting of headers is a bit harder since method calls don't have direct access to headers. The workaround is either using `AxiosRequestTransformer` or defining the header as a parameter for all endpoints.
7. 🟡 Endpoint descriptions have the method of `'OPTIONS'` removed.
8. 🔴 Plugins package has not yet migrated to v11. The `tokenPlugin` still works 🤞.
9. 🔴 The `isErrorFromAlias` and `isErrorFromPath` don't work with Beta 19 version for endpoints with path params https://github.com/ecyrbe/zodios/pull/237#issuecomment-2003188263


@ecyrbe keep up the good work 👏. Any info on another beta release or v11 stable release?

**Reply by jordanmnunez** [reply]
+1 on all these feedback @toteto. Excited to hear an update as well.

**Comment by corneliusroemer** [comment]
Are there still plans for v11?
